### ISSUE TITLE
Errors are obscured when interacting with multiple Turbines

### ISSUE THREAD

**Issue by jingibus** [cashapp/turbine#165]
In the following test:

 ```kotlin
  @Test fun failingTest() = runTest {
    val otherTurbine = Turbine<String>()
    flow<Nothing> {
      throw RuntimeException("I barfed!")
    }.test {
      assertEquals("next item", otherTurbine.awaitItem())
    }
  }
```

...the programmer never sees the `RuntimeException` failure. They only see the assertion error from `otherTurbine`.

This can be really, really confusing when the `RuntimeException` is the root cause of the problem.

**Comment by JakeWharton** [comment]
This is tricky to design correctly. The exception thrown from the test method could either be from this Turbine (event type mismatch), general user code (assertion error), or from another Turbine (which is kinda the same as user code).

This would be a lot easier if we had suppressed exceptions like we do on the JVM.

I have three ideas for things we can do to potentially clear these cases up:
- For exceptions originating from Turbine (such as event type mismatch), include a summary of the remaining items and complete/error states.
- For exceptions originating from a test lambda, determine whether they originated from the Turbine receiver or not, and if not then wrap the exception with one that includes that same summary of remaining items.
- For standalone Turbines inside a scope, use a coroutine context element to register them to some single instance which can aggregate a summary of all unconsumed events from all Turbines when the scope is existing due to failure or just normally.

**Comment by JakeWharton** [comment]
Also tricky is how to make sure these don't compound on each other as the exception propagates.

**Comment by jingibus** [comment]
This problem has been bugging the hell out of me. Because in e.g. this scenario:

```kotlin
val componentA = FakeA()
val flowB = flowToTest(componentA)
val turbineB = flowB.testIn(this)

assertThat(componentA.outputTurbine.awaitItem()).isEqualTo(55)
```

....we do not have any way to get at `turbineB`'s contents if that last assertion fails for any reason. 

We can solve this problem by forcing a turbine scope if one is not already in a `test` validation block:

```kotlin
turbineScope {
  val componentA = FakeA()
  val flowB = flowToTest(componentA)
  val turbineB = flowB.testIn(this)

  assertThat(componentA.outputTurbine.awaitItem()).isEqualTo(55)
}
```

We'd have to enforce this dynamically in `testIn`, because I don't think there's a way right now to make the type system do it for us. (Unless context receivers are usable rn?)

I think that would avoid the compounding issue: `turbineScope` annotates with its bound turbines as exceptions bubble up, and each turbine is only bound to one `turbineScope`

Standalone Turbines could either:

A. Skip registration entirely, you're on your own, or
B. Register opportunistically when an `await` method is called

**Comment by JakeWharton** [comment]
Oh yeah I kinda forgot Turbine's can just be out there in the wild. Hell, you could be `take`-ing from multiple in a fully synchronous test and we have no means of correlating them.

Definitely a lot to experiment with here.

**Comment by jingibus** [comment]
They can just be out there in the wild, yes, but note that when they are, we aren't responsible for eating their exceptions. Whoever is using them is.

Having said that, internally at Cash we have `Turbines` and `TurbinesRule` that call `ensureAllEventsConsumed()` at the end of your test. Something like `turbineScope` could do this in a strictly better way, reporting *all* the unconsumed turbine events, without a rule hookup. But that's a pretty big behavior change.

**Comment by matejdro** [comment]
It appears that, multiple turbines are not even necessary. Following code passes, even though it should fail, since there were actual events (an exception):

```kotlin
val flowThatThrows = flow<Nothing> {
   throw IllegalStateException()
}

flowThatThrows.test {
   expectNoEvents()
}
```

This is with turbine 0.12.1. 

Should I create a new issue instead or is this related?

**Comment by jingibus** [comment]
I think that can get its own issue. Please include the entire failing test, btw, and platform — it is helpful to know which dispatcher we're looking at.

**Comment by JakeWharton** [comment]
Note we have a test which is exactly that setup: https://github.com/cashapp/turbine/blob/1e115ae8c8c46435bf9e9c6be717bc0fcf18c63b/src/commonTest/kotlin/app/cash/turbine/FlowTest.kt#L167-L180 (the end of the `test` block implicitly calls `expectNoEvents()`).

So I suspect you likely have additional operators between the flow and the call to `.test` which use dispatchers and thus make the error's propagation through the stream non-synchronous.

**Comment by jingibus** [comment]
This issue has been on my mind again after a colleague came to me with a stumper that ended up being caused by this issue. It is extremely baffling when it happens!

So I've been having shower thoughts about how this might go. This shape is kind of the pencil mark I've been thinking with:

```kotlin
turbineScope {
  val componentA = FakeA()
  val flowB = flowToTest(componentA)
  val turbineB = flowB.testIn(this)

  assertThat(componentA.outputTurbine.awaitItem()).isEqualTo(55)
}
```

That would at least delimit when we start and stop looking for exceptions in Turbines: at the start of `turbineScope` would be when we start looking for Turbine exceptions, and at the end is when we stop and maybe report them.

Before getting to the question of what and how to implement this `turbineScope`, though, there's a more important question: which exceptions do we actually want to report?  Let's pencil in the following scope and think through it:

> *_Any exception in any Turbine_*

Off the top of my head I can think of a couple of scenarios covered by that that would really expand the frameworkiness of Turbine in a way I don't like:

* _Exceptions in standalone Turbines._ Doing this exceeds the bounds of what `turbineScope` can accomplish: a standalone Turbine may be constructed outside of `turbineScope`, and could be completed exceptionally at any time, not just within that block scope. So for us to report all of these, we have no other choice than to hook this new machinery in to the Turbine at either its construction site or at the exceptional completion site.
* _Exceptions thrown in non-coroutine tests._ `take` should only be called in non-coroutine tests (or at least, they aren't intended to be supported outside that situation). That means that `turbineScope` would have to have a non-suspending variant, which would not be able to use something sensible like a context element. 

So these are a pain to support. Are they even important to support?

I don't think so. None of the stumpers I've been sent that relate to this issue were related to either of those two scenarios. They are in fact pretty well covered by the things you could imagine `turbineScope` covering in a structured concurrent manner.

So here's another scope:

> *_Any exception thrown in a coroutine or block scope being run by Turbine._*

This makes a lot of sense to me? It would mean that this behavior is only meaningful in `.test { ... }` and `.testIn(...)`. Standalone turbines wouldn't be affected at all, but that's fine: you would have to manually add your own exception there, which is your own responsibility. Maybe some API would be needed in the future there for people to thread in their own exceptions to standalone Turbines in a tracked way, but I can't currently imagine any way that would be useful beyond what `testIn` provides. It may never come up; if it does, it can get built then.

Apart from all the above, the only other thorny thing is actually reporting the additional exceptions. But that's much less thorny than the corner cases above IMO.

Sound sensible?

**Comment by jingibus** [comment]
I got a ways down the road implementing this, and ran into a corner: `backgroundScope`.

Let's say we support throwing in this scenario:

```kotlin
runTest {
  flowOf(1, 2, 3).test {
    val subturbine = flow { throw RuntimeException() }.testIn(this)

    assertThat(awaitItem()).isEqualTo(1)
  }
} 
```

This we can do by wrapping `validate` in a `withContext` and adding our own elements to the `CoroutineContext`. But this:

```kotlin
runTest {
  flowOf(1, 2, 3).test {
    val subturbine = flow { throw RuntimeException() }.testIn(backgroundScope)

    assertThat(awaitItem()).isEqualTo(1)
  }
} 
```

...we can't do without adding a lot of additional API surface, because `TestScope` is sealed.

I think there is a sensible way out, though, which is to alter the semantics of `testIn` to match the semantics of `test`. Previously, that was not a guarantee that Turbine could make; if Turbine requires either a `withTurbines { ... }` or `.test { ... }`, though, it can. That will include an implicit cancellation, which would obviate the need for `backgroundScope`, which would mean Turbine can just not support it.

**Comment by jingibus** [comment]
Addressed this for a lot of scenarios with #233 .

Looking at two additional changes to address this:

1. An additional `turbine { ... }` API that provides the same additional support we added to `.test { ... }` for folks who are only using testIn.
2. Breaking API change: if `testIn` is called without a `TurbineRegistryElement`, throw.

I think (1) is a no-brainer.

I think (2), even though it's a breaking change so close to 1.0, is pretty close to a no-brainer: even if we include a `turbine { ... }` API, if we allow usage of `testIn` without `turbine { ... }  or `.test { ... }`, it is like adding a head lamp to the library, but putting it back home in a junk drawer instead of in their backpack. Anyone who reaches for it is most certainly already committed, and will have to backport it all over the place to get back to safety. Strictness will result in a lot less pain.
